### PR TITLE
[Enhancement] push down aggregate on grouping-set

### PR DESCRIPTION
**Pull request by Seaven**
## What I'm doing:
grouping set sql
```
     select a, b, c, d, e sum(f) from t group by rollup(a, b, c, d, e);
```

rewrite to:
```
   with cte1 as (select a, b, c, d, e, sum(f) x from t group by a, b, c, d, e -- shuffle by a high cardinality column, for better effect on local aggregate
     selec * fom cte1
     union all
     select a, b, c, d, null, sum(x) x from cte1 group by rollup(a, b, c, d)
```

if the last column in rollup/cube/groupingset：
* is a high cardinality column,  it's has a significant performance improvement. (like tpcds q67)
* is a low cardinality column,  it's has no effect.

vs `set enable_rewrite_groupingsets_to_union_all = true` (will rewite whole grouping set to CTE)
* save CTE copy cost
* I think has no any particular disadvantages, maybe repeat node will use more memory?

tpcds Q67 100g: 13.492s -> 10.253s
tpcds Q67 1t: 29.06s -> 10s

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownAggregateGroupingSetsRule.java [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+        super(RuleType.TF_PUSHDOWN_AGG_GROUPING_SET,
+                Pattern.create(OperatorType.LOGICAL_AGGR)
+                        .addChildren(Pattern.create(OperatorType.LOGICAL_REPEAT, OperatorType.PATTERN_LEAF)));
+    }
+
+    @Override
+    public boolean check(OptExpression input, OptimizerContext context) {
+        LogicalAggregationOperator aggregate = (LogicalAggregationOperator) input.getOp();
+        LogicalRepeatOperator repeatOperator = (LogicalRepeatOperator) input.inputAt(0).getOp();
+        if (aggregate.getType() != AggType.GLOBAL || repeatOperator.getRepeatColumnRef().size() <= 3
```
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownAggregateGroupingSetsRule.java:73]: If there are too few rollup columns, I think the performace impove isn't significant.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownAggregateGroupingSetsRule.java [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+            Utils.calculateStatistics(input, context);
+        }
+        if (null != repeatInput.getStatistics()) {
+            // use one column to shuffle
+            Statistics statistics = repeatInput.getStatistics();
+            partitionRefs = allGroupByRefs.stream()
+                    .filter(ref -> !statistics.getColumnStatistic(ref).isUnknown())
+                    .sorted((o1, o2) -> Double.compare(statistics.getColumnStatistic(o2).getDistinctValuesCount(),
+                            statistics.getColumnStatistic(o1).getDistinctValuesCount()))
+                    .limit(1)
```
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownAggregateGroupingSetsRule.java:163]: for make the local aggregate has better aggregation result which one after repeat node

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownAggregateGroupingSetsRule.java [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+
+        LogicalCTEConsumeOperator consume = new LogicalCTEConsumeOperator(cteId, cteColumnRefs);
+
+        // project
+        int lastGroups = repeat.getRepeatColumnRef().size() - 1;
+        for (int i = 0; i < repeat.getOutputGrouping().size(); i++) {
+            ColumnRefOperator input = repeat.getOutputGrouping().get(i);
+            ColumnRefOperator output = factory.create(input, input.getType(), input.isNullable());
+            outputs.put(input, output);
+            projectMap.put(output, ConstantOperator.createBigint(repeat.getGroupingIds().get(i).get(lastGroups)));
```
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownAggregateGroupingSetsRule.java:224]: I added

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java [hunk 4]
```diff
@@ -595,6 +594,15 @@ private OptExpression logicalRuleRewrite(
...
     }
 
+    private void rewriteGroupingSets(OptExpression tree, TaskContext rootTaskContext, SessionVariable sessionVariable) {
+        if (sessionVariable.isCboPushDownGroupingSet()) {
+            ruleRewriteOnlyOnce(tree, rootTaskContext, new PushDownAggregateGroupingSetsRule());
+        }
+        if (sessionVariable.isEnableRewriteGroupingsetsToUnionAll()) {
+            ruleRewriteIterative(tree, rootTaskContext, new RewriteGroupingSetsByCTERule());
+        }
+    }
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:604]: If both rules are turn on, will it split to union twice?

**Review by LiShuMing — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java (hunk 4 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:604]: emmm, you are right, will execute two rules and split all to cte in final, may be we need execute `RewriteGroupingSetsByCTERule` frist

**Review by packy92 — APPROVED** [review]

**Review by LiShuMing — APPROVED** [review]

**Comment by Youngwb** [comment]
@Mergify backport branch-3.3

**Pull request merged**